File: src/main/java/com/sber/yamlprocessor/model/Subprocess.java
package com.sber.yamlprocessor.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;

import java.util.List;

public class Subprocess {
    
    private String id;
    
    @JsonProperty("context-code")
    private ContextCodes contextCode;
    
    private String description;
    
    private Boolean disabled = false;
    
    @Valid
    @NotNull(message = "Trigger is required")
    private Trigger trigger;
    
    @Valid
    @NotNull(message = "Stages is required")
    private List<Stage> stages;

    public Subprocess() {}

    public Subprocess(String description, Trigger trigger, List<Stage> stages) {
        this.description = description;
        this.trigger = trigger;
        this.stages = stages;
    }

    // Getters and Setters
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public ContextCodes getContextCode() {
        return contextCode;
    }

    public void setContextCode(ContextCodes contextCode) {
        this.contextCode = contextCode;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Boolean getDisabled() {
        return disabled;
    }

    public void setDisabled(Boolean disabled) {
        this.disabled = disabled;
    }

    public Trigger getTrigger() {
        return trigger;
    }

    public void setTrigger(Trigger trigger) {
        this.trigger = trigger;
    }

    public List<Stage> getStages() {
        return stages;
    }

    public void setStages(List<Stage> stages) {
        this.stages = stages;
    }
}

File: src/main/kotlin/com/sber/yamlprocessor/model/Subprocess.kt
package com.sber.yamlprocessor.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Embedded
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.OrderColumn
import jakarta.persistence.Table

@Entity
@Table(name = "subprocess")
class Subprocess(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonIgnore
    @Column(name = "id")
    var dbId: Long? = null,

    @ManyToOne(optional = false)
    @JoinColumn(name = "process_id")
    @JsonIgnore
    var process: Process? = null,

    @Column(name = "business_id")
    var id: String? = null,

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "context_code", referencedColumnName = "code")
    @JsonProperty("context-code")
    var contextCode: ContextCodesDictionary? = null,

    @Column(name = "description", nullable = false, length = 4000)
    var description: String = "",

    @Column(name = "disabled")
    var disabled: Boolean = false,

    @Embedded
    var trigger: Trigger = Trigger(),

    @OneToMany(mappedBy = "subprocess", cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderColumn(name = "stage_sort")
    var stages: MutableList<Stage> = mutableListOf()
)
